### ISSUE TITLE
[Design] Deferred invocation for actors ("Entities" model)

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/dapr#7122]
## In what area(s)?

/area runtime

## Describe the feature

This is a design for addressing the issue in #6842

It is not a "new feature" in Dapr, but rather a new API for exposing an existing capability in a much more ergonomic way.

### Business problem

Original issue is #6842

Users want to be able to "fire and forget" an actor. That is: they want to be able to invoke an actor asynchronously, without making the caller wait for the response. A message is added to the actor's queue, and then it's executed once the actor is able to. (The caller does not receive a response)

This is similar to the ["Entities" model](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-entities?tabs=function-based%2Cin-process&pivots=csharp)

### How this works in Dapr today

This pattern is already possible in Dapr today. In fact, that's how the Dapr Workflow engine works under the hood!

However, using this pattern with Dapr is not something we document, and even the underlying primitives aren't documented clearly. That is, it requires having an understanding of some of Dapr's internals.

To be able to schedule a deferred message:

- A user needs to create an actor reminder with a "0s" execution time.
- Ideally, the reminder should also have a repetition (for example every 1 min), in order to have the message be retried automatically if it failed. Once the reminder is executed successfully, the caller returns a response indicating Dapr to drop the repeating message.
- The current APIs implemented by Dapr SDKs normally allow only actors to register reminders for themselves. However, the Dapr APIs allow anyone to schedule a reminder, as long as the sidecar receiving the request is capable of hosting reminders of that kind (ideally, the registration should happen in the sidecar that is hosting the given actor). This means users should route the request to the actor host first.

### Proposed new API

In order to make this capability easier to use for users, I propose the creation of a new API:

```proto
service Dapr {
  rpc InvokeActorDeferred (InvokeActorDeferredRequest) returns (InvokeActorDeferredResponse) {}
}

// InvokeActorDeferredRequest is the request for InvokeActorDeferred.
message InvokeActorDeferredRequest {
  // These properties are also present in InvokeActorRequest
  string actor_type = 1;
  string actor_id = 2;
  string method = 3;
  bytes data = 4;
  map<string, string> metadata = 5;

  // Timeout for the execution of the message
  google.protobuf.Duration deadline = 6;
}

// InvokeActorDeferredResponse is the response from InvokeActorDeferred.
message InvokeActorDeferredResponse {
  // Empty
}
```

This request can be made to any Dapr actor **client** (it requires the Dapr runtime to have an active actor client at least). When the message is received:

- If the current runtime is the host of the given actor (actor type or actor ID):
  - The sidecar creates a new reminder for the given actor, with the requested actor type and ID, and a random name.
  - The reminder has "0s" execution time and a repeating time of 1 minute.
  - The reminder's data contains the requested method, the metadata, the deadline, and the data passed by the user _(note: we know that today our reminders subsystem doesn't scale well when reminders have a lot of data; we should document using caution if passing a lot of data)_
- If the current runtime isn't the host of the given actor:
  - Dapr will make a "service invocation" call (using a new internal gRPC method `ScheduleActorReminder`) to the sidecar that hosts the given actor. This call is made synchronously and the caller (the app that wants to do the deferred invocation) waits for it to return.
  - When the message arrives in the sidecar that is the host for the given actor, the steps above are performed.
- When the reminder is executed in the actor host:
  - Dapr invokes the actor in background once it's not busy. The request has a timeout equal to the deadline set in the request.
  - The response from the app is discarded, but the status code of the response is logged and reported as metric.
  - When the reminder is completed, Dapr deletes the repeating reminder.
  - If the execution fails (e.g. if the Dapr sidecar crashes), the reminder is eventually repeated until successful.
  - On the other hand, if the app returns a non-retriable error, the reminder is dropped. The status code of the response is logged, and the operation is reported as metric.

## Release Note

RELEASE NOTE: **ADD** Deferred invocation API for actors

**Comment by cgillum** [comment]
I have a few thoughts on this but take them with a grain of salt since I don't represent the typical user of Dapr Actors.

**One important clarification first**: what's being proposed here isn't really the "entities" model as defined by Azure Durable Functions, but something closer to the _traditional_ actor model (before RPC-style virtual actors were introduced). I just want to clarify that up front in case people arrive at this issue thinking that it proposes the actual Durable Entities programming model, which is quite a bit more than asynchronous actor invocation (I'd be happy to explain in more details if that would be useful but will refrain for now).

With that out of the way, a few comments I'd like to contribute:

> In fact, that's how the Dapr Workflow engine works under the hood

I actually considered implementing this type of pattern for Dapr Workflow originally. However, I abandoned that plan for a couple reasons:

1. There were cases where I wanted to validate the message and potentially return an error before saving the reminder, which wouldn't be possible if I scheduled the reminder directly from the caller.
2. I was very much concerned about the performance impact of data payloads inside of reminders. I see that you rightfully called that out in this proposal, though I understand this problem is more of an implementation detail related to the reminder subsystem.

I figured it might be worth mentioning this since, although it addresses a similar problem we had when building the workflow backend, it's not a solution I think we'd actually use given the points above.

> Ideally, the reminder should also have a repetition (for example every 1 min), in order to have the message be retried automatically if it failed. Once the reminder is executed successfully, the caller returns a response indicating Dapr to drop the repeating message.

I wonder if this might actually be a separate feature - i.e., _at-least-once_ reminder invocations. As I understand things right now, actor reminders are essentially _at-most-once_. This could be a useful feature even without asynchronous actor invocation.

> A message is added to the actor's **queue**, and then it's executed once the actor is able to. (The caller does not receive a response)

This design is essentially a mechanism for turning the actor reminder subsystem into a queuing mechanism. That's not necessarily a bad thing, but I do wonder: why not instead do proper Actor pub/sub integration as proposed in https://github.com/dapr/dapr/issues/501? It seems to me that this has the potential to address the core problems in a more natural way, but with a host of other benefits that come with using pub/sub in Dapr (and I'd _love_ to plug that feature into Dapr Workflows as it would enable some really cool scenarios that not even Durable Functions can do). Have you considered this as an alternative?

**Comment by ItalyPaleAle** [comment]
Thanks @cgillum  for the review!

> One important clarification first: what's being proposed here isn't really the "entities" model as defined by Azure Durable Functions, but something closer to the traditional actor model (before RPC-style virtual actors were introduced)

I have removed mentions of entities to prevent confusion.

> I actually considered implementing this type of pattern for Dapr Workflow originally. However, I abandoned that plan for a couple reasons:
> 
> 1. There were cases where I wanted to validate the message and potentially return an error before saving the reminder, which wouldn't be possible if I scheduled the reminder directly from the caller.
> 2. I was very much concerned about the performance impact of data payloads inside of reminders. I see that you rightfully called that out in this proposal, though I understand this problem is more of an implementation detail related to the reminder subsystem.

For WF, these seem very valid point. I think 1 is a special case for WF.

For 2, I agree that is a concern. I am writing this design with the assumption that the scaling issues on reminders are going to be fixed (they are being worked on). Even if they weren't fixed fully, users could store large payloads out-of-band, for example in a state store, and put in the reminder just a "pointer" to the data in the state store.

> I wonder if this might actually be a separate feature - i.e., at-least-once reminder invocations. As I understand things right now, actor reminders are essentially at-most-once. This could be a useful feature even without asynchronous actor invocation.

I think so, we should look into that (separately from this issue)

> This design is essentially a mechanism for turning the actor reminder subsystem into a queuing mechanism. That's not necessarily a bad thing, but I do wonder: why not instead do proper Actor pub/sub integration as proposed in https://github.com/dapr/dapr/issues/501? It seems to me that this has the potential to address the core problems in a more natural way, but with a host of other benefits that come with using pub/sub in Dapr (and I'd love to plug that feature into Dapr Workflows as it would enable some really cool scenarios that not even Durable Functions can do). Have you considered this as an alternative?

I think #501 is a very good issue that we should, one day, tackle. I don't consider this proposal as an exact alternative to that, however.

To me, deferred messaging are more like "fire and forget" background operations. For example, they could be done for garbage collection purposes.

There are similarities with a queue, but also differences. The biggest one, of course, is that this scenario does not depend on a separate queue service, for good and bad reasons (no need to maintain an external service, but also no interactions with external systems/callers).

**Comment by olitomlinson** [comment]
Great to see thought going into this! From what I can tell, the community would really value this!

> what's being proposed here isn't really the "entities" model as defined by Azure Durable Functions, but something closer to the traditional actor model (before RPC-style virtual actors were introduced)

I agree with Chris. @ItalyPaleAle I would consider revising your original post/title and removing the concept of Entities here as to not cause confusion.

---

I don't have much to add to the implementation details but I would recommend bringing this back to the DevEx.

**As a** Dev
**I want** the process for 'Fire-and-forget' to be as similar to the existing `invocation` method.
**So that** it keeps the SDK API surface simple, consistent, and swapping between RPC and Fire-And-Forget is trivial

**Comment by yaron2** [comment]
#501 is about removing boilerplate code developers need to write today to receive a pub/sub event and then invoking the desired actor via RPC, essentially routing code from pub/sub event to actor invocation. So I do not think it correlates much to what's being proposed here.

**Comment by olitomlinson** [comment]
Agree with @yaron2  - Also, I see this as a technical pre-req to https://github.com/dapr/dapr/issues/501 as Actor PubSub can only scale well if callers are not blocked.

**Comment by cgillum** [comment]
I don't want to necessarily push for this, but I don't understand why #501 doesn't solve the core problem that raised this issue? Pub/sub is fundamentally a means of doing async messaging, and if it's plugged into actors, then actors now have async, fire-and-forget invocation support. I also don't understand why the current proposal as written a technical pre-req for actor pub/sub. Am I misunderstanding something, either this proposal or the Actor pub/sub proposal?

**Comment by olitomlinson** [comment]
[#501 ](https://github.com/dapr/dapr/issues/501)comes with lots of baggage, ranging from fan-out of a single published message to all instances of one or more Actor Types, to a single published message which targets a specific Actor instance. None of that is related to this particular ticket.

PubSub brokers scale best when they are unencumbered by subscribers that indefinitely block, as it stands the only API surface to 'invoke' Actors is via synchronous RPC which _does_ block. So IMO this needs addressing first (via this ticket, for example),in order to make for a good scalable Actor PubSub implementation later on down the line.

Of course I'm not an authority on the matter, so I could be very wrong :)

**Comment by ItalyPaleAle** [comment]
I was just talking to @artursouza about whether we could have stateless actors. As a reminder, the Orleans whitepaper does include stateless actors which have no limit on concurrency (as they are meant to be stateless).

I didn’t think they were making much sense in dapr since we have service invocation, BUT they can make sense if we use deferred invocation for actors.

even better, having stateless actors (with no concurrency limits) means that this feature can even be implemented via a built-in, stateless actor, so it doesn’t require adding new APIs to the internal service. The stateless actor would be of the same type as the actor for which the reminder is created.

**Comment by olitomlinson** [comment]
@ItalyPaleAle 

Have I got this right...

Introduce a _stateless_ actor as an internal 'inbox' that receives a **single** invocation request, and then relays the request onto the target stateful actor?

In the example below I've tried to demonstrate a single actor (actor X) making **two** "fire and forget" requests to a target actor (actor y)

- Each "fire and forget" call will be received by an ephemeral _stateless_ inbox actor. The stateless inbox actor simply Acks back to the caller (actor x). 
- The stateless inbox actor will then retry indefinitely to relay the request to the target actor (actor y)

<img width="794" alt="image" src="https://github.com/dapr/dapr/assets/4224880/af9ef486-c0c6-4bbc-a67c-b708b88db44b">

**Comment by ItalyPaleAle** [comment]
@olitomlinson not quite.

The problem we need to solve is that reminders (at least in the current implementation in Dapr) can only be scheduled by a sidecar that is a host capable of hosting actors of the given type. For example, if you have app-a and app-b, and app-b can host actors of type "myactor", then calls to register reminders on those actors must go to "app-b". (I am not 100% sure what happens if you schedule a reminder on a host that is not the host for that actor instance - if you have your app scaled horizontally, that is - will need to be tested).

My proposal above (which is still very much viable) requires creating a new RPC in the Dapr Internal gRPC service (the one used for sidecar-to-sidecar communication).

If we can use stateless actors (which do not have concurrency limits), then we could use those instead **to schedule a reminder**.

1. app-b hosts actors of type "myactor"
1. app-a wants to do a deferred invocation on "myactor/id1"
2. app-a does an invocation for a stateless actor "myactor/<random ID>" (we may need to do some tweaks to make sure that if app-b is scaled horizontally, "myactor/<random ID>" is placed on the same host as "myactor/id1" if that's necessary - see my question above)
3. "myactor/<random ID>" receives the invocation and schedules the reminder, then responds to "app-a" synchronously
4. The reminder is now scheduled with 0s repetitions, and that can be used to do the deferred invocation (and the invocation can be on a stateful or stateless actor)

**Comment by olitomlinson** [comment]
@ItalyPaleAle ahhh I see now! Thanks for clarifying!

**Comment by ItalyPaleAle** [comment]
👋🤖

**Comment by Manuel-Moya** [comment]
It would be great having the fire and forget functionality for dapr actors, I would love to move from AKKA to Dapr once that is in place.
